alf/algorithms/handcrafted_algorithm.py

```diff
@@ -0,0 +1,195 @@
+                batched environment, which means that it runs multiple simulations
+                simultateously. ``env` only needs to be provided to the root
+                algorithm.
+            config (TrainerConfig): config for training. It only needs to be
+                provided to the algorithm which performs ``train_iter()`` by
+                itself.
+            debug_summaries (bool): True if debug summaries should be created.
+            name (str): The name of this algorithm.
+        """
+
+        super().__init__(
+            observation_spec,
+            action_spec,
+            env=env,
+            config=config,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        self._distance_to_decelerate = distance_to_decelerate
+        self._distance_to_stop = distance_to_stop
+
+    def _policy_func(self, observation):
+        """A naive hand-crafted policy for Carla environment.
+
+        Args:
+            observation (nested Tensor): input observation that is compatible
+                with observation_spec
+        Returns
+            Tensor: action computed based on the observation
+        """
+        # waypoints is a [B, k, 3] shaped tensor, which contains the batched
+        # positions of a number of future waypoints in the route, relative to
+        # the coordinate system of the respective vehicle. waypoints[:, 0]
+        # is the closest waypoint and waypoints[:, -1] is the farthest one.
+        # Each waypoint has 3 elements corresponding to the x, y, z values
+        # relative to the vehicle's coordinate system.
+        waypoints = alf.nest.get_field(observation, 'observation.navigation')
+
+        # goal is a [B, 3] shaped tensor, with each 3D vector contains the
+        # x, y, z positions of the goal, relative to  to the vehicle's
+        # coordinate system.
+        goal = alf.nest.get_field(observation, 'observation.goal')
+
+        if waypoints.shape[1] > 1:
+            wp_vector = waypoints[:, 1]
+        else:
+            wp_vector = waypoints[:, 0]
+
+        direction = torch.atan2(wp_vector[..., 1], wp_vector[..., 0])
+        direction = direction / np.pi
+
+        # action is a [B, 3] tensor with each 3D vector corresponding to
+        # [speed, direction, reverse].
+        # speed: 1.0 corresponding to maximally allowed speed
+        # direction: relative to the vehicle's heading, with 0 being front,
+        # -0.5 being left and 0.5 being right
+        # reverse: values greater than 0.5 corrsponding to going backward.
+
+        action = torch.zeros(waypoints.shape[0], self._action_spec.shape[0])
+
+        distance_to_goal = torch.norm(goal, dim=1)
+
+        # here we adjust the speed based on the distance to goal
+        action[distance_to_goal > self._distance_to_decelerate, 0] = 1
+        # logical_and is introduced in torch 1.5.0, use * instead
+        ind = (distance_to_goal > self._distance_to_stop) * (
```

Comment:
FYI: & can be used

Reply:
Good suggestion. I have removed the comment and changed to &